File: ray/ir/inst.py
from __future__ import annotations
from typing import List
from ..Ast import Literal, Ty, BinaryKind, Lit
from enum import Enum, auto
from .basic_block import BasicBlock


class Value:
    pass


class Inst(Value):
    _id = 0

    def __init__(self, name=None) -> None:
        if not name:
            self.inst_id = Inst._id
            Inst._id += 1
        self.name = name
        self.parent: BasicBlock
        self.ty: Ty

    @classmethod
    def reset(cls):
        cls._id = 0

    def uses(self, i: Inst) -> bool:
        raise NotImplemented()

    def __str__(self) -> str:
        return "%{}".format(self.inst_id if not self.name else self.name)

    def __repr__(self) -> str:
        return "%{}".format(self.inst_id if not self.name else self.name)


class Alloc(Inst):
    def __init__(self, ty: Ty, off: int, name: str) -> None:
        super().__init__()
        self.ty = ty
        self.size = ty.get_size()
        self.off = off
        self.var_name = name
        self.userz: List[Inst] = []

    def users(self, basic_block: BasicBlock) -> List[Inst]:
        users = set()
        for inst in basic_block.instructions:
            if inst.uses(self):
                users.add(inst)
        return list(users)

    def remove_from_parent(self):
        self.parent.instructions.remove(self)

    def uses(self, _: Inst) -> bool:
        return False

    def __str__(self) -> str:
        return "    {} = alloc {}".format(super().__str__(), self.ty)


class Store(Inst):
    def __init__(self, dst: Value, src: Value, name: str) -> None:
        super().__init__()
        self.dst = dst
        self.src = src
        self.var_name = name

    def uses(self, i: Inst) -> bool:
        match self.src:
            case IConst():
                return self.dst == i
            case _:
                return self.dst == i or self.src == i

    def __str__(self) -> str:
        return "    store {}, {}".format(repr(self.dst), repr(self.src))


class Load(Inst):
    def __init__(self, src: Inst) -> None:
        super().__init__()
        self.src = src

    def uses(self, i: Inst) -> bool:
        return self.src == i

    def __str__(self) -> str:
        return "    {} = load {}".format(super().__str__(), repr(self.src))


class CmpKind(Enum):
    Lt = auto()
    Gt = auto()
    Eq = auto()
    NotEq = auto()

    @staticmethod
    def from_binary(kind: BinaryKind):
        match kind:
            case BinaryKind.Lt:
                return CmpKind.Lt
            case BinaryKind.Gt:
                return CmpKind.Gt
            case BinaryKind.Eq:
                return CmpKind.Eq
            case BinaryKind.NotEq:
                return CmpKind.NotEq
            case _:
                assert False, kind

    def __repr__(self) -> str:
        match self:
            case CmpKind.Lt:
                return "lt"
            case CmpKind.Gt:
                return "gt"
            case CmpKind.Eq:
                return "eq"
            case CmpKind.NotEq:
                return "noteq"


class Cmp(Inst):
    def __init__(self, left: Value, right: Value, kind: CmpKind) -> None:
        super().__init__()
        self.left = left
        self.right = right
        self.kind = kind

    def uses(self, i: Inst) -> bool:
        return self.left == i or self.right == i

    def __str__(self) -> str:
        return "    {} = cmp {} {}, {}".format(super().__str__(), repr(self.kind), repr(self.left), repr(self.right))


class Bin(Inst):
    def __init__(self, left: Value, right: Value) -> None:
        super().__init__()
        self.left = left
        self.right = right

    def uses(self, i: Inst) -> bool:
        return self.left == i or self.right == i

    def __str__(self) -> str:
        return "    {} = {} {}, {}".format(super().__str__(), self.inst_name, repr(self.left), repr(self.right))


class Add(Bin):
    def __init__(self, left: Value, right: Value) -> None:
        super().__init__(left, right)
        self.inst_name = "add"


class Sub(Bin):
    def __init__(self, left: Value, right: Value) -> None:
        super().__init__(left, right)
        self.inst_name = "sub"


class Mul(Bin):
    def __init__(self, left: Value, right: Value) -> None:
        super().__init__(left, right)
        self.inst_name = "mul"


class Div(Bin):
    def __init__(self, left: Value, right: Value) -> None:
        super().__init__(left, right)
        self.inst_name = "div"


class Mod(Bin):
    def __init__(self, left: Value, right: Value) -> None:
        super().__init__(left, right)
        self.inst_name = "mod"


class Br(Inst):
    __match_args__ = ('cond', 'btrue', 'bfalse', )

    def __init__(self, cond: Value, btrue: int, bfalse: int) -> None:
        super().__init__()
        self.cond = cond
        self.btrue = btrue
        self.bfalse = bfalse

    def uses(self, i: Inst):
        return self.cond == i

    def __str__(self) -> str:
        return "    br {}, %bb{}, %bb{}".format(repr(self.cond), self.btrue, self.bfalse)


class Jmp(Inst):
    __match_args__ = ('br_id', )

    def __init__(self, br_id: int) -> None:
        super().__init__()
        self.br_id = br_id

    def uses(self, _: Inst) -> bool:
        return False

    def __str__(self) -> str:
        return "    jmp %bb{}".format(self.br_id)


class Phi(Inst):
    def __init__(self, btrue: Value, bfalse: Value) -> None:
        super().__init__()
        self.btrue = btrue
        self.bfalse = bfalse

    def uses(self, _: Inst) -> bool:
        return False

    def __str__(self) -> str:
        return "    {} = phi {}, {}".format(super().__str__(), self.btrue, self.bfalse)


class Nop(Inst):
    def __init__(self) -> None:
        super().__init__()

    def uses(self, _: Inst) -> bool:
        return False

    def __str__(self) -> str:
        return "    nop"


class ConstKind(Enum):
    Int = auto()
    Float = auto()
    Str = auto()
    Bool = auto()


class IConst(Value):
    def __init__(self, kind, value) -> None:
        self.kind = kind
        self.value = value

    @staticmethod
    def from_lit(lit: Literal) -> IConst:
        match lit.kind:
            case Lit.Int:
                return IConst(ConstKind.Int, lit.value)
            case Lit.Float:
                return IConst(ConstKind.Float, lit.value)
            case Lit.Str:
                return IConst(ConstKind.Str, lit.value)
            case Lit.Bool:
                return IConst(ConstKind.Bool, lit.value)
            case _:
                assert False, lit.kind

    def __str__(self) -> str:
        return self.value

    __repr__ = __str__

File: ray/ir/inst.ml
open Middle.Ty
open Printf
open Structures.Vec

type t = {
    kind: inst_kind
  ; mutable id: int
}

and binary_kind =
  | Add
  | Sub
  | Mul
  | Div
  | Gt
  | Lt
  | Eq
  | NotEq
  | GtEq
  | LtEq
  | BitAnd
  | BitOr
  | And
  | Or

and inst_kind =
  | Alloca of ty ref
  | Binary of binary_kind * value * value
  | Br of value * value * value
  | Jmp of value
  (* (bb * inst) *)
  | Phi of ty * (value * value) list
  | Store of value * value
  | Load of value
  | Gep of ty * value * int
  | Call of ty * value * value list
  | Intrinsic of string * value list
  | Ret of value
  | Trap of value
  (* Bitcast of ty * value *)
  (* Zext of ty * value *)
  (* Sext of ty * value *)
  (* Trunc of ty * value *)
  (* Fpext of ty * value *)
  (* Fptrunc of ty * value *)
  | PtrToInt of value * ty
  | IntToPtr of value * ty
  | RetUnit
  | Nop

and value =
  | Const of const * ty ref
  | VReg of t * ty ref
  | Label of basic_block
  | Param of ty * string * int
  | Global of string

and const =
  | Int of int
  | Float of float
  | Str of string
  | Bool of bool
  | Struct of value list

and basic_block = {
    mutable pred: basic_block vec
  ; mutable succ: basic_block vec
  ; mutable insts: t vec
  ; mutable bid: int
  ; mutable is_entry: bool
}
(* [@@deriving show] *)

let binary_kind_to_inst = function
  | Ast.Add -> Add
  | Sub -> Sub
  | Mul -> Mul
  | Div -> Div
  | Gt -> Gt
  | Lt -> Lt
  | Eq -> Eq
  | NotEq -> NotEq
  | GtEq -> GtEq
  | LtEq -> LtEq
  | BitAnd | And -> And
  | BitOr | Or -> Or
;;

let render_binary = function
  | Add -> "add"
  | Sub -> "sub"
  | Mul -> "mul"
  | Div -> "div"
  | Gt -> "gt"
  | Lt -> "lt"
  | Eq -> "eq"
  | NotEq -> "neq"
  | GtEq -> "gte"
  | LtEq -> "lte"
  | BitAnd | And -> "and"
  | BitOr | Or -> "or"
;;

let has_value = function
  | Br _ | Jmp _ | Store _ | Ret _ | RetUnit | Nop -> false
  | _ -> true
;;

let rec render_const = function
  | Int value -> sprintf "%d" value
  | Float value -> sprintf "%f" value
  | Str value -> sprintf "%s" (String.escaped value)
  | Bool value -> sprintf "%b" value
  | Struct values ->
      sprintf "{ %s }" (String.concat ", " (List.map render_value values))

and render_value = function
  | Const (const, ty) -> sprintf "%s %s" (render_ty !ty) (render_const const)
  | VReg (inst, ty) -> sprintf "%s%%%i" (render_ty !ty ^ " ") inst.id
  | Label bb -> sprintf "label %%bb%d" bb.bid
  | Param (ty, name, _) -> sprintf "%s %%%s" (render_ty ty) name
  | Global name -> sprintf "@%s" name
;;

let value_with_ty value ty =
  match value with
  | Const (const, _) -> Const (const, ref ty)
  | VReg (kind, _) -> VReg (kind, ref ty)
  | Param (_, name, i) -> Param (ty, name, i)
  | Global _ | Label _ -> value
;;

let get_ty = function
  | Param (ty, _, _) -> ref ty
  | Const (_, ty) | VReg (_, ty) -> ty
  | _ -> assert false
;;

let extract_block = function Label bb -> bb | _ -> assert false

let render_inst inst : string =
  (if has_value inst.kind then sprintf "    %%%d = " inst.id else "    ")
  ^
  match inst.kind with
  | Alloca ty -> sprintf "alloca %s" (render_ty !ty)
  | Binary (kind, left, right) ->
      sprintf
        "%s %s, %s"
        (render_binary kind)
        (render_value left)
        (render_value right)
  | Br (cond, true_block, false_block) ->
      sprintf
        "br %s, %s, %s"
        (render_value cond)
        (render_value true_block)
        (render_value false_block)
  | Phi (ty, args) ->
      sprintf
        "phi %s, %s"
        (render_ty ty)
        (String.concat
           ", "
           (List.map
              (fun (bb, inst) ->
                sprintf "[%s, %s]" (render_value bb) (render_value inst))
              args))
  | Jmp bb -> sprintf "jmp %s" (render_value bb)
  | Store (src, dst) ->
      sprintf "store %s, %s" (render_value src) (render_value dst)
  | Load ptr ->
      sprintf
        "load %s, %s"
        (match !(get_ty ptr) with
         | Ptr ty -> render_ty ty
         | Ref ty -> render_ty ty
         | _ -> assert false)
        (render_value ptr)
  | Gep (_, value, index) -> sprintf "gep %s, %d" (render_value value) index
  | Call (ty, fn, args) ->
      let ty = match ty with FnPtr { ret; _ } -> ret | _ -> assert false in
      sprintf
        "call %s %s(%s)"
        (render_ty ty)
        (render_value fn)
        (String.concat ", " (List.map render_value args))
  | Intrinsic (name, args) ->
      sprintf
        "intrinsic %s(%s)"
        name
        (String.concat ", " (List.map render_value args))
  | PtrToInt (value, ty) ->
      sprintf "ptrtoint %s to %s" (render_value value) (render_ty ty)
  | IntToPtr (value, ty) ->
      sprintf "inttoptr %s to %s" (render_value value) (render_ty ty)
  | Ret ret -> sprintf "ret %s" (render_value ret)
  | Trap _ -> "trap"
  | RetUnit -> "ret"
  | Nop -> "nop"
;;
